swim/heal_partition_test.go

```diff
@@ -0,0 +1,344 @@
+
+	targets := A[0].node.healer.Heal()
+	assert.Len(t, targets, 1, "expected correct amount of targets")
+	assert.True(t, B.Contains(targets[0]), "expected target to be a node from the right partition")
+
+	waitForPartitionHeal(t, time.Second*3, A, B)
+
+	A.HasPartitionAs(t, A, 0, "alive")
+	A.HasPartitionAs(t, B, 0, "alive")
+	B.HasPartitionAs(t, B, 0, "alive")
+	B.HasPartitionAs(t, A, 0, "alive")
+}
+
+func TestPartitionHealWithFaultyAndMissing(t *testing.T) {
+	A := newPartition(t, 6)
+	A1 := partition(A[0:3])
+	defer destroyNodes(A...)
+
+	B := newPartition(t, 6)
+	B1 := partition(B[0:3])
+	defer destroyNodes(B...)
+
+	A.ProgressTime(time.Millisecond * 3)
+	B.ProgressTime(time.Millisecond * 5)
+
+	// A.AddPartitionWithStatus(B1, Suspect)
+	A.AddPartitionWithStatus(B1, Faulty)
+	// B.AddPartitionWithStatus(A1, Suspect)
+	B.AddPartitionWithStatus(A1, Faulty)
+
+	waitForConvergence(t, 3*time.Second, A...)
+	waitForConvergence(t, 3*time.Second, B...)
+
+	A[0].node.discoverProvider = statichosts.New(append(A.Hosts(), B.Hosts()...)...)
+	A[0].node.healer.Heal()
+	waitForConvergence(t, 3*time.Second, A...)
+	waitForConvergence(t, 3*time.Second, B...)
+
+	A[0].node.healer.Heal()
+	waitForPartitionHeal(t, time.Second*3, A, B)
+}
+
+// TestPartitionHealSemiPartition checks if the ring cluster automatically
+// recovers when half partitioned. The bidirectional full sync mechanism
+// is responsible to heal partitions of this form. We test for this because
+// the partition healing mechanism could leave the cluster in this state
+// when it completes only partially.
+func TestPartitionHealSemiParition(t *testing.T) {
+	A := newPartition(t, 5)
+	B := newPartition(t, 5)
+	defer destroyNodes(A...)
+	defer destroyNodes(B...)
+
+	A.AddPartitionWithStatus(B, Alive)
+
+	waitForPartitionHeal(t, time.Second*3, A, B)
+}
+
+func TestPartitionHealWithTime(t *testing.T) {
+	A := newPartition(t, 5)
+	B := newPartition(t, 5)
+	defer destroyNodes(A...)
+	defer destroyNodes(B...)
+
+	A.AddPartitionWithStatus(B, Faulty)
+	B.AddPartitionWithStatus(A, Faulty)
+
+	A.ProgressTime(time.Millisecond)
+	B.ProgressTime(time.Millisecond)
+
+	A[0].node.discoverProvider = statichosts.New(append(A.Hosts(), B.Hosts()...)...)
+
+	A[0].node.healer.Start()
+
+	// Progress time in a background thread. This will cause the node to
+	// attempt a heal periodically.
+	c := clock.NewMock()
+	A[0].node.clock = c
+	go func() {
+		for {
+			c.Add(time.Second)
+		}
+	}()
+
+	waitForPartitionHeal(t, 3000*time.Millisecond, A, B)
+}
+
+// TestHealBeforeBootstrap tests that we can heal with a node that is still
+// bootstrapping. We put a node in the bootstrapping phase by giving it a
+// host-list that is impossible to join. When it is in this phase, we trigger
+// a heal attempt from another node. The heal should succeed from b's
+// perspective. Healing from bootstrap is very similar to joining a node that
+// is in the bootstrapping phase.
```

Comment:
The last sentence is confusing. How is that similar to join? To what degree? If you are comparing, it has to be much more detailed (or avoid comparing altogether).

But, still, "very similar" means "a bit different". How is it different?